### PR TITLE
Class library resources

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
This should allow localization to find resources in a class library by using an Assembly attribute to note the location of the resources folder.

### CODE REVIEW

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 1]
```diff
@@ -65,12 +66,27 @@ public IStringLocalizer Create(Type resourceSource)
...
+            var resourcePath = _resourcesRelativePath;
+            var baseNamespace = _hostingEnvironment.ApplicationName;
+
+            //This isn't the main assembly
+            if (!string.Equals(assemblyName.Name, _hostingEnvironment.ApplicationName, StringComparison.Ordinal))
+            {
+                var resourceLocationAttribute = assembly.GetCustomAttribute<ResourceLocationAttribute>();
+                // If we don't have an attribute, just assume it's in 'Resources'
+                resourcePath = string.IsNullOrEmpty(resourceLocationAttribute?.ResourceFolder)
+                    ? "Resources"
```
Comment (ryanbrandenburg) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: After thinking about this, I think the right thing to do here, rather than just assuming the folder is "Resources", is to do basically what we do for project resources and leave it null so we get TypeInfo.FullName.

**Comment by ryanbrandenburg** [comment]
@javiercn I believe you're the next most knowledgeable for localization since @dougbu is unavailable.

**Comment by hishamco** [comment]
Copyright & docs are missing in the source files

File: src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Localization
+{
+    public class ResourceLocationAttribute : Attribute
+    {
+        public ResourceLocationAttribute(string resourceFolder)
+        {
+            ResourceFolder = resourceFolder;
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs]: Why we didn't throw `ArgumentNullException` instead of assuming the resource folder name is "Resources"

File: src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        }
+
+        public string ResourceFolder { get; private set; }
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs]: I think readonly property in here is preferable

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 1 above)
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: This will be modify to
```csharp
resourcePath = resourceLocationAttribute?.ResourceFolder;
```
if we throw the exception above in the `ResourceLocationAttribute`

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 4]
```diff
@@ -96,11 +112,23 @@ public IStringLocalizer Create(string baseName, string location)
...
+                var assembly = Assembly.Load(assemblyName);
+
+                var resourceLocation = _resourcesRelativePath;
+
+                // in another assembly
+                if (!string.Equals(assemblyName.Name, _hostingEnvironment.ApplicationName, StringComparison.Ordinal))
+                {
+                    var resourcesLocationAttribute = assembly.GetCustomAttribute<ResourceLocationAttribute>();
+                    resourceLocation = !string.IsNullOrEmpty(resourcesLocationAttribute?.ResourceFolder)
+                            ? resourceLocation = resourcesLocationAttribute.ResourceFolder + "."
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Again we can avoid the `string.IsNullOrEmpty` check everywhere

**Comment by dougbu** [comment]
@ryanbrandenburg I've read some related comments elsewhere in this repo or perhaps Home or MVC. Could you link to the relevant issues or comments to help me w/ the details? In particular what is the problem this design PR is aimed to solve?

**Comment by hishamco** [comment]
@dougbu AFAIK this design related to the issues https://github.com/aspnet/Localization/issues/281#issuecomment-247173762, https://github.com/aspnet/Localization/issues/275#issuecomment-247174158

File: test/ResourcesClassLibraryNoAttribute/project.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿{
+  "version": "1.0.0-*",
+  "buildOptions": {
+    "warningsAsErrors": true,
+    "keyFile": "../../tools/Key.snk"
+  },
+  "dependencies": {
+    "NETStandard.Library": "1.6.0"
```
Comment (dougbu) [inline test/ResourcesClassLibraryNoAttribute/project.json]: Why not use a lower .NET standard library?

File: test/Microsoft.AspNetCore.Localization.FunctionalTests/LocalizationTest.cs [hunk 6]
```diff
@@ -12,7 +12,27 @@ namespace Microsoft.AspNetCore.Localization.FunctionalTests
     public class LocalizationTest
     {
         private static readonly string _applicationPath = Path.Combine("test", "LocalizationWebsite");
-        
+
+        [ConditionalTheory]
+        [OSSkipCondition(OperatingSystems.Linux)]
+        [OSSkipCondition(OperatingSystems.MacOSX)]
```
Comment (ryanbrandenburg) [inline test/Microsoft.AspNetCore.Localization.FunctionalTests/LocalizationTest.cs]: So far as I can tell the idea is to avoid running CLR on Non-windows. But there seems to be a lot of duplication, I can file an issue to clean all of the tests in this format up at once.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: test/Microsoft.AspNetCore.Localization.FunctionalTests/LocalizationTest.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Localization.FunctionalTests/LocalizationTest.cs]: plz

File: src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+                throw new ArgumentNullException(nameof(resourceLocation));
+            }
+
+            ResourceLocation = resourceLocation;
+        }
+
+        /// <summary>
+        /// The location of the resources for this Assembly
+        /// </summary>
+        public readonly string ResourceLocation;
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs]: Readonly property not a ```readonly``` field
```csharp
public string ResourceLocation { get; }
```

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 8]
```diff
@@ -170,6 +174,20 @@ public IStringLocalizer Create(string baseName, string location)
...
         }
 
+        protected virtual string GetResourcePrefix(string location, string baseName, string resourceLocation)
+        {
+            // Re-root the base name if a resources path is set
+            return location + "." + resourceLocation + TrimPrefix(baseName, location + ".");
+        }
+        private string GetResourcePath(Assembly assembly)
+        {
+            var resourceLocationAttribute = assembly.GetCustomAttribute<ResourceLocationAttribute>();
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: To allow unit testing, need a method or two which has a `ResourceLocationAttribute` parameter and is available from test assembly. For example, confirm correct resource path is chosen if `ResourceLocation` contains directory separators.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 9]
```diff
@@ -170,6 +174,20 @@ public IStringLocalizer Create(string baseName, string location)
...
+            // If we don't have an attribute assume all assemblies use the same resource location.
+            return string.IsNullOrEmpty(resourceLocationAttribute?.ResourceLocation)
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: `ResourceLocation` can't be `null` if the attribute exists. Just check `resourceLocationAttribute == null`.

File: test/LocalizationWebsite/StartupResourcesInFolder.cs [hunk 10]
```diff
@@ -43,6 +44,8 @@ public void ConfigureServices(IServiceCollection services)
             });
 
             var stringLocalizer = stringLocalizerFactory.Create("Test", location: null);
+            var assembly = typeof(StartupResourcesInFolder).GetTypeInfo().Assembly;
+            var stringLocalizerExpliciteLocation = stringLocalizerFactory.Create("Test", assembly.FullName);
```
Comment (dougbu) [inline test/LocalizationWebsite/StartupResourcesInFolder.cs]: Frequente typoe in "Explicit"

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 11]
```diff
@@ -109,5 +109,26 @@ public void Create_FromNameLocation_ReturnsNewResultForDifferentLocation()
...
+            var locOptions = new LocalizationOptions();
+            var options = new Mock<IOptions<LocalizationOptions>>();
+            options.Setup(o => o.Value).Returns(locOptions);
+            var factory = new ResourceManagerStringLocalizerFactory(hostingEnvironment.Object, localizationOptions: options.Object);
+            var location1 = typeof(ResourceManagerStringLocalizer).GetTypeInfo().Assembly.FullName;
+            var location2 = typeof(ResourceManagerStringLocalizerFactoryTest).GetTypeInfo().Assembly.FullName;
+
+            // Act
+            var result1 = factory.Create("baseName", null);
+            var result2 = factory.Create("baseName", null);
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Names for the `null` arguments

File: src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Localization
+{
+    /// <summary>
+    /// Gives the location of the resources for an Assembly.
+    /// </summary>
+    public class ResourceLocationAttribute : Attribute
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs]: Be explicit about the `[AttributeUsage]`

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 13]
```diff
@@ -157,11 +159,13 @@ public IStringLocalizer Create(string baseName, string location)
 
             location = location ?? _hostingEnvironment.ApplicationName;
 
-            baseName = GetResourcePrefix(baseName, location);
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Not using this `protected virtual` method is a breaking change. Those who have overridden it will no longer see their preferred behaviour.

File: src/Microsoft.Extensions.Localization/project.json [hunk 14]
```diff
@@ -22,7 +22,8 @@
     "Microsoft.AspNetCore.Hosting.Abstractions": "1.1.0-*",
     "Microsoft.Extensions.DependencyInjection.Abstractions": "1.1.0-*",
     "Microsoft.Extensions.Localization.Abstractions": "1.1.0-*",
-    "Microsoft.Extensions.Options": "1.1.0-*"
+    "Microsoft.Extensions.Options": "1.1.0-*",
+    "System.Reflection.Extensions": "4.0.1"
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/project.json]: `"4.0.1-*"`

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+using Microsoft.Extensions.Localization;
+using Microsoft.Extensions.Logging;
+
+namespace LocalizationWebsite
+{
+    public class StartupResourcesInClassLibrary
+    {
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddLocalization(options => options.ResourcesPath = "Resources");
```
Comment (dougbu) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: Separate from work in this design PR but do we thoroughly test use of `ResourcesPath` e.g. values containing directory separators?

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+            var noAttributeName = new AssemblyName(noAttributeAssembly.FullName).Name;
+            var noAttributeNameStringLocalizer = stringLocalizerFactory.Create(
+                nameof(ResourcesClassLibraryNoAttribute.Model),
+                noAttributeName);
+
+            var withAttributeAssembly = typeof(ResourcesClassLibraryWithAttribute.Model).GetTypeInfo().Assembly;
+            var withAttributeName = new AssemblyName(withAttributeAssembly.FullName).Name;
+            var withAttributeNameStringLocalizer = stringLocalizerFactory.Create(
+                nameof(ResourcesClassLibraryWithAttribute.Model),
+                withAttributeName);
```
Comment (dougbu) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: 1. Only testing one `Create()` overload.
2. Is all this boilerplate really needed? Doesn't look particularly usable.

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 17]
```diff
@@ -99,8 +99,8 @@ public void Create_FromNameLocation_ReturnsNewResultForDifferentLocation()
             var options = new Mock<IOptions<LocalizationOptions>>();
             options.Setup(o => o.Value).Returns(locOptions);
             var factory = new ResourceManagerStringLocalizerFactory(hostingEnvironment.Object, localizationOptions: options.Object);
-            var location1 = typeof(ResourceManagerStringLocalizer).GetTypeInfo().Assembly.FullName;
-            var location2 = typeof(ResourceManagerStringLocalizerFactoryTest).GetTypeInfo().Assembly.FullName;
+            var location1 = new AssemblyName(typeof(ResourceManagerStringLocalizer).GetTypeInfo().Assembly.FullName).Name;
+            var location2 = new AssemblyName(typeof(ResourceManagerStringLocalizerFactoryTest).GetTypeInfo().Assembly.FullName).Name;
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Why change this test? Have the semantics w/o a `[ResourceLocation]` changed?

Comment (ryanbrandenburg) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: The test gave the impression that we should be passing in the full assemblyName to `Create(string, string)` which would make the ResourceManager look for something like `TestAssembly, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e.Resources.TestClass`. We actually want to pass the "short" assembly name there. But we didn't notice because the Assert in that test only cares if it produces Same/NotSame factories.

TL:DR: The semantics haven't changed, this was just wrong to start with.

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs (hunk 16 above)
Comment (ryanbrandenburg) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: I agree that `IStringLocalizerFactory.Create(string, string)` requires an awful lot of boilerplate, and is confusing as compared to `IStringLocalizerFactory.Create(Type)`. I would recommend very heavily that people use the Type overload, but I don't know if we're to the point of deprecating it, and I don't know of any way to make it simpler to use.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Localization
+{
+    /// <summary>
+    /// Gives the location of the resources for an Assembly.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = false)]
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs]: Be explicit about `Inherited` choice as well.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 19]
```diff
@@ -62,7 +62,7 @@ protected virtual string GetResourcePrefix(TypeInfo typeInfo)
                 throw new ArgumentNullException(nameof(typeInfo));
             }
 
-            return GetResourcePrefix(typeInfo, _hostingEnvironment.ApplicationName, _resourcesRelativePath);
+            return GetResourcePrefix(typeInfo, new AssemblyName(typeInfo.Assembly.FullName).Name, _resourcesRelativePath);
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: This change won't matter (`GetResourcePrefix()` will still return `typeInfo.FullName` which is a strange resource prefix) when `_resourcesRelativePath` is `null` or empty. If that's a possible scenario, should think the semantics through a bit more. If it's not possible, remove the ternary at [line 91](https://github.com/aspnet/Localization/pull/283/files#diff-ec3f46a4394df7156f745d976f053127R91).

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 20]
```diff
@@ -155,21 +159,50 @@ public IStringLocalizer Create(string baseName, string location)
...
+                _resourceNamesCache);
+        }
+
+        protected virtual string GetResourcePrefix(string location, string baseName, string resourceLocation)
+        {
+            // Re-root the base name if a resources path is set
+            return location + "." + resourceLocation + TrimPrefix(baseName, location + ".");
+        }
+
+        protected virtual ResourceLocationAttribute GetResourceLocationAttribute(Assembly assembly)
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Not sure about the other `protected virtual` methods but please add a comment identifying this one as `virtual` for testing. Can the other two be `private` or (better) `private static`?

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs (hunk 15 above)
Comment (dougbu) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: ??

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 21]
```diff
@@ -1,17 +1,102 @@
...
 
-namespace Microsoft.Extensions.Localization.Test
+namespace Microsoft.Extensions.Localization.Tests
 {
+    public class TestResourceManagerStringLocalizerFactory : ResourceManagerStringLocalizerFactory
+    {
+        private ResourceLocationAttribute _resourceLocationAttribute;
+
+        public Assembly Assembly { get; set; }
+        public string BaseName { get; set; }
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Curious: Why are these `public`ly settable?

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 22]
```diff
@@ -109,5 +217,48 @@ public void Create_FromNameLocation_ReturnsNewResultForDifferentLocation()
             // Assert
             Assert.NotSame(result1, result2);
         }
+
+        [Fact]
+        public void Create_FromNameLocation_ResourcesPathDirectorySeperatorToDot()
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Plze spele separator correctly

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 23]
```diff
@@ -109,5 +217,48 @@ public void Create_FromNameLocation_ReturnsNewResultForDifferentLocation()
...
+            var factory = new TestResourceManagerStringLocalizerFactory(
+                hostingEnvironment.Object,
+                options.Object,
+                resourceLocationAttribute: null);
+
+            // Act
+            var result1 = factory.Create("baseName", location: null);
+
+            // Assert
+            Assert.DoesNotContain(Path.DirectorySeparatorChar, factory.BaseName);
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Look for the actual expected resource path to match the test name. Should not be OS-sensitive.

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 24]
```diff
@@ -109,5 +217,48 @@ public void Create_FromNameLocation_ReturnsNewResultForDifferentLocation()
...
+        {
+            // Arrange
+            var hostingEnvironment = new Mock<IHostingEnvironment>();
+            hostingEnvironment.SetupGet(a => a.ApplicationName).Returns("Microsoft.Extensions.Localization.Tests");
+            var locOptions = new LocalizationOptions();
+            var options = new Mock<IOptions<LocalizationOptions>>();
+            options.Setup(o => o.Value).Returns(locOptions);
+            var factory = new ResourceManagerStringLocalizerFactory(hostingEnvironment.Object, localizationOptions: options.Object);
+            var location1 = typeof(ResourceManagerStringLocalizer).GetTypeInfo().Assembly.FullName;
+            var location2 = typeof(ResourceManagerStringLocalizerFactoryTest).GetTypeInfo().Assembly.FullName;
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: What are `locationX` for?

File: test/ResourcesClassLibraryNoAttribute/project.json (hunk 5 above)
Comment (dougbu) [inline test/ResourcesClassLibraryNoAttribute/project.json]: ??

**Comment by dougbu** [comment]
:watch: for you to declare this non-design. Looks close to me but double-check the doc comments and extra / missing blank lines.

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs (hunk 15 above)
Comment (ryanbrandenburg) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: We do not, which I know because it was broken. I added Create_FromType_ResourcesPathDirectorySeperatorToDot and Create_FromNameLocation_ResourcesPathDirectorySeperatorToDot to cover this area.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 19 above)
Comment (ryanbrandenburg) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: It is possible and correct for GetResourcePrefix to return typeInfo.FullName. In the case that we don't have a Resource path it would return something like `Microsoft.Extensions.Localization.Tests.ResourceManagerStringLocalizerFactoryTest` which is how it will be stored in the assembly. Or are you asking something else?

File: test/ResourcesClassLibraryNoAttribute/project.json (hunk 5 above)
Comment (ryanbrandenburg) [inline test/ResourcesClassLibraryNoAttribute/project.json]: Because if I do I get a warning that it ended up with NETStandard.Library 1.6.0 anyway. What are the downside of using 1.6 considering that this a a test library?

Comment (dougbu) [inline test/ResourcesClassLibraryNoAttribute/project.json]: :ok: guess I was mostly curious why the project needed this particular .NET standard.

File: src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs [hunk 25]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) .NET Foundation. All rights reserved. 
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Localization
+{
+    /// <summary>
+    /// Gives the location of the resources for an Assembly.
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs]: Gives -> Provides and nuke the second "the".

Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceLocationAttribute.cs]: Same "the" removal twice below.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs (hunk 19 above)
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: :ok:

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 26]
```diff
@@ -155,21 +157,70 @@ public IStringLocalizer Create(string baseName, string location)
...
+                baseName = GetResourcePrefix(baseName, location);
+
+                return CreateIStringLocalizer(assembly, baseName);
             });
         }
 
+        /// <summary>
+        /// This method is virtual for testing purposes only.
+        /// </summary>
+        /// <param name="assembly">The assembly to create an IStringLocalizer for.</param>
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: `<see cref="IStringLocalizer"/>`

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 27]
```diff
@@ -155,21 +157,70 @@ public IStringLocalizer Create(string baseName, string location)
...
+        /// <param name="baseName">The base name of the resource to search for.</param>
+        /// <returns></returns>
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Add a blurb here and in the two empty `<returns>` elements below.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 28]
```diff
@@ -155,21 +157,70 @@ public IStringLocalizer Create(string baseName, string location)
...
+                var assemblyName = new AssemblyName(location);
+                var assembly = Assembly.Load(assemblyName);
+                baseName = GetResourcePrefix(baseName, location);
+
+                return CreateIStringLocalizer(assembly, baseName);
             });
         }
 
+        /// <summary>
+        /// This method is virtual for testing purposes only.
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: A fine `<remarks>` but what does this method _do_?

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs [hunk 29]
```diff
@@ -155,21 +157,70 @@ public IStringLocalizer Create(string baseName, string location)
...
+        /// <returns></returns>
+        protected virtual ResourceLocationAttribute GetResourceLocationAttribute(Assembly assembly)
+        {
+            return assembly.GetCustomAttribute<ResourceLocationAttribute>();
+        }
+
+        private string GetResourcePath(Assembly assembly)
+        {
+            var resourceLocationAttribute = GetResourceLocationAttribute(assembly);
+            // If we don't have an attribute assume all assemblies use the same resource location.
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs]: Blank line above to improve readability.

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs [hunk 30]
```diff
@@ -0,0 +1,70 @@
+﻿using System;
```
Comment (dougbu) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: ©

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs (hunk 16 above)
Comment (dougbu) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: Where is `IStringLocalizerFactory.Create(Type)` tested?

File: test/ResourcesClassLibraryNoAttribute/Properties/AssemblyInfo.cs [hunk 31]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Reflection;
```
Comment (dougbu) [inline test/ResourcesClassLibraryNoAttribute/Properties/AssemblyInfo.cs]: :copyright:

File: test/ResourcesClassLibraryWithAttribute/Properties/AssemblyInfo.cs [hunk 32]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Reflection;
```
Comment (dougbu) [inline test/ResourcesClassLibraryWithAttribute/Properties/AssemblyInfo.cs]: :copyright:

**Comment by dougbu** [comment]
:watch: for some doc comment cleanup

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs (hunk 16 above)
Comment (ryanbrandenburg) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: We create `IStringLocalizer`'s with that overload above this on L43 and L44, then use them on L62 and L66. They also have unit tests in ResourceManagerStringLocalizerFactoryTest.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: test/LocalizationWebsite/StartupResourcesInClassLibrary.cs (hunk 16 above)
Comment (dougbu) [inline test/LocalizationWebsite/StartupResourcesInClassLibrary.cs]: :ok_hand:

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs [hunk 33]
```diff
@@ -1,17 +1,79 @@
...
 
+using System.IO;
 using System.Reflection;
 using Microsoft.AspNetCore.Hosting;
 using Microsoft.Extensions.Options;
 using Moq;
 using Xunit;
 
-namespace Microsoft.Extensions.Localization.Test
+namespace Microsoft.Extensions.Localization.Tests
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: nit: Shouldn't this be in the `Microsoft.Extensions.Localization` namespace?

Comment (ryanbrandenburg) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Every other file in this repo uses adds Tests to the namespace it's testing, so if that is the standard, we're breaking it uniformly here.

Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerFactoryTest.cs]: Right, we're breaking the [guidelines](https://github.com/aspnet/Home/wiki/Engineering-guidelines#unit-test-class-naming) uniformly:
> Test class names end with Test and live in the same namespace as the class being tested.

But you're also right that this is a separate issue from this PR.
